File: examples/control/SMC/SMCExampleImplicit.cpp
/* Siconos is a program dedicated to modeling, simulation and control
 * of non smooth dynamical systems.
 *
 * Copyright 2023 INRIA.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* !\file SMCExampleImplicit.cpp
  \brief Two independent systems of dimension one controlled to slide
  on \f$x = 0\f$. An implicit scheme is used
  O. Huber
  */

#include <SiconosControl.hpp>
#include <SiconosKernel.hpp>
#include <chrono>
#include <string>
using Matrix = siconos::algebra::SiconosMatrix;
using Vector = siconos::algebra::SiconosVector;
using namespace std;

class MyDS : public siconos::modeling::FirstOrderLinearDS {
 public:
  MyDS(auto x0, auto A) : FirstOrderLinearDS(x0, A) {
    _b = std::make_shared<Vector>(x0->size());
  };
  void computeb(double time) {
    // printf("computeB\n");
    double t = sin(50 * time);
    _b->setValue(0, t);
    _b->setValue(1, -t);
    // printf("b[0] = %g, b[1] = %g\n", _b->getValue(0), _b->getValue(1));
  };
};
// main program
int main(int argc, char* argv[]) {
  // User-defined parameters
  unsigned int ndof = 2;     // Number of degrees of freedom of your system
  double t0 = 0.0;           // Starting time
  double T = 1;              // Total simulation time
  double h = 1.0e-4;         // Time step for simulation
  double hControl = 1.0e-2;  // Time step for control
  double Xinit = 1.0;

  if (h > hControl) {
    THROW_EXCEPTION("hControl must be bigger than h");
  }

  // ================= Creation of the model =======================
  // Steps:
  // - create a Dynamical System
  // - add a Simulation to the model

  // -------------------------
  // --- Dynamical systems ---
  // -------------------------

  // First System:
  // dx/dt = Ax + u(t) + r
  // x(0) = x0
  // Note: r = Blambda, B defines in relation below.

  // Matrix declaration
  auto A = std::make_shared<Matrix>(ndof, ndof, 0);
  auto x0 = std::make_shared<Vector>(ndof);
  (*x0)(0) = Xinit;
  (*x0)(1) = -Xinit;
  auto sensorC = std::make_shared<Matrix>(2, 2);
  sensorC->setIdentity();
  auto sensorD = std::make_shared<Matrix>(2, 2, 0);
  auto Csurface = std::make_shared<Matrix>(1, 2, 0);
  (*Csurface)(0, 1) = 1;
  auto Brel = std::make_shared<Matrix>(2, 1, 0);
  (*Brel)(1, 0) = 2;

  // Dynamical Systems
  auto processDS = std::make_shared<MyDS>(x0, A);
  // -------------
  // --- Model process ---
  // -------------
  auto sim = std::make_shared<siconos::control::ControlZOHSimulation>(t0, T, h);
  // sim->setSaveOnlyMainSimulation(true);
  sim->addDynamicalSystem(processDS);

  // ------------------
  // --- Simulation ---
  // ------------------
  // Control stuff
  // use a controlSensor
  auto sens = std::make_shared<siconos::control::LinearSensor>(processDS, sensorC, sensorD);
  sim->addSensor(sens, hControl);
  // add the sliding mode controller
  auto act = std::make_shared<siconos::control::LinearSMC>(sens);
  act->setCsurface(Csurface);
  act->setB(Brel);
  sim->addActuator(act, hControl);
  // =========================== End of model definition ===========================

  // ================================= Computation =================================

  // --- Simulation initialization ---

  cout << "====> Simulation initialisation ...\n\n";
  // initialise the process and the ControlManager
  sim->initialize();

  // ==== Simulation loop =====
  cout << "====> Start computation ... \n\n";
  sim->run();
  // --- Output files ---
  cout << "====> Output file writing ...\n";
  auto& dataPlot = *sim->data();

  siconos::algebra::io::write("SMCExampleImplicit.dat", dataPlot,
                              siconos::algebra::io::ASCII_OUT,
                              siconos::algebra::io::WriteType::nodim);

  double error = 0.0, eps = 1e-12;
  if ((error = siconos::algebra::io::compareRefFile(dataPlot, "SMCExampleImplicit.ref", eps)) >
      eps)
    return 1;
  else
    return 0;

  // // Comparison with a reference file
  // SiconosMatrix dataPlotRef(dataPlot);
  // dataPlotRef.setZero();
  // ioMatrix::read("SMCExampleImplicit.ref", "ascii", dataPlotRef);
  // std::cout << (dataPlot - dataPlotRef).normInf() << std::endl;

  // if ((dataPlot - dataPlotRef).normInf() > 1e-12)
  // {
  //   std::cout << "Warning. The results is rather different from the reference file." <<
  //   std::endl; return 1;
  // }
}

File: examples/control/SMC/SMCExampleImplicit.py
#!/usr/bin/env python

# Siconos is a program dedicated to modeling, simulation and control
# of non smooth dynamical systems.
#
# Copyright 2021 INRIA.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from siconos.kernel import FirstOrderLinearDS
from siconos.control.simulation import ControlZOHSimulation
from siconos.control.sensor import LinearSensor
from siconos.control.controller import LinearSMC

from numpy import eye, zeros, savetxt
from math import ceil
from matplotlib import rc
import scipy
from scipy import arange

import distutils.spawn

if distutils.spawn.find_executable("latex"):
    rc("text", usetex=True)
import siconos.plot_config as sicoplot

# Turn off interactive backend by default
plt, enable_plot = sicoplot.choose_backend(False)

# variable declaration
ndof = 2  # Number of degrees of freedom of your system
t0 = 0.0  # start time
T = 1  # end time
h = 1.0e-4  # time step for simulation
hControl = 1.0e-2  # time step for control
Xinit = 1.0  # initial position
theta = 0.5
N = int(ceil((T - t0) / h + 10))  # number of time steps
outputSize = 5  # number of variable to store at each time step

# Matrix declaration
A = zeros((ndof, ndof))
x0 = [Xinit, -Xinit]
sensorC = eye(ndof)
Csurface = [[0, 1]]
Brel = [[0], [2]]
# Drel = [[0, 0]]
# Simple check
if h > hControl:
    print("hControl must be bigger than h")
    exit(1)

# Declaration of the Dynamical System
processDS = FirstOrderLinearDS(x0, A)
processDS.setComputebFunction("RelayPlugin", "computeB")

# Control simulation
sim = ControlZOHSimulation(t0, T, h)
sim.setSaveOnlyMainSimulation(True)
sim.addDynamicalSystem(processDS)
# Actuator, Sensor & ControlManager
sens = LinearSensor(processDS, sensorC)
sim.addSensor(sens, hControl)
act = LinearSMC(sens)
act.setCsurface(Csurface)
act.setB(Brel)
sim.addActuator(act, hControl)

# Initialization
sim.initialize()

# Run simulation
sim.run()

# Get data
dataPlot = sim.data()

# Save to disk
savetxt("SMCExampleImplicit-py.dat", dataPlot)
# Plot interesting data

plt.subplot(211)
plt.ylabel(r"$\sigma$")
plt.xlabel(r"t")
plt.plot(dataPlot[:, 0], dataPlot[:, 2])
plt.grid()
plt.subplot(212)
plt.ylabel(r"$\bar{u}^s$")
plt.xlabel(r"t")
plt.ylim(-2.1, 2.1)
plt.plot(dataPlot[:, 0], dataPlot[:, 3])
plt.savefig("ismc_sigma_u.png")

plt.subplot(211)
plt.ylabel(r"$\sigma$")
plt.xlabel(r"t")
plt.xlim(xmin=0.49)
plt.ylim(-0.03, 0.03)
plt.plot(dataPlot[4900:, 0], dataPlot[4900:, 2])
plt.grid()
plt.subplot(212)
plt.ylabel(r"$\bar{u}^s$")
plt.xlabel(r"t")
plt.ylim(-2.1, 2.1)
plt.xlim(xmin=0.49)
p1 = plt.plot(dataPlot[4900:, 0], dataPlot[4900:, 3])
# p2 = plt.plot(dataPlot[4900:, 0], np.sin(50*dataPlot[4900:, 0]))
# plt.legend((p1[0], p2[0]), (r'$\bar{u}^s(t)$', r'$-\rho(t)$'), ncol=2)
plt.savefig("ismc_sigma_u_z.png")

u_z = dataPlot[5100:, 3]
n = len(u_z)
Y = scipy.fft(dataPlot[5100:, 3]) / n
k = arange(n)
T = n * h
frq = k / T
frq = frq[list(range(int(n / 2)))]
Y = Y[list(range(int(n / 2)))]
plt.plot(frq, abs(Y), "r")
plt.xlabel(r"freq (Hz)")
plt.title(r"Frequency spectrum of $\bar{u}^s$")
plt.savefig("ismc_u_freq.png")

# TODO
# compare with the reference
# ref = getMatrix(SiconosMatrix("result.ref"))
# if (norm(dataPlot - ref[1:,:]) > 1e-12):
#    print("Warning. The result is rather different from the reference file.")
